_2020/shell-tools.md

```diff
@@ -198,7 +198,7 @@ All UNIX-like systems come packaged with [`find`](https://www.man7.org/linux/man
 # Find all directories named src
 find . -name src -type d
 # Find all python files that have a folder named test in their path
-find . -path '**/test/**/*.py' -type f
+find . -path '*/test/*/*.py' -type f
```

Comment:
For the `find` commands, I'm not sure if the lecture uses a non-standard version of `find`, but GNU `find` doesn't recognize `**` as a shell pattern for the `-path` command, and `*` explicitly matches `/` (see [manual](https://www.gnu.org/software/findutils/manual/html_mono/find.html#Shell-Pattern-Matching)). And `{.}` below expands to itself in GNU `find`.

Reply:
I actually think this isn't quite right (and neither was the original version in the text).

Consider the following directory structure:

```
.
├── test
│   ├── d.py
│   ├── test
│   │   └── a.py
│   └── x
│       └── b.py
└── x
    └── test
        └── c.py
```

```console
$ find . -path '*/test/*/*.py' -type f
./test/test/a.py
./test/x/b.py
```

It doesn't find "all python files that have a folder named test in their path".

Reply:
Oh, you're right! I'll update to this, so it matches the behaviour described in the comment (and the behaviour of `find . -path '**/test/**/*.py' -type f` if it behaved like Bash with the `globstar` shell option):

```console
$ find . -path '*/test/*.py' -type f
./x/test/c.py
./test/x/b.py
./test/d.py
./test/test/a.py
```